### ISSUE TITLE
Function call with wrong number of arguments

### ISSUE THREAD

**Issue by rain-1** [biwascheme/biwascheme#129]
Trying to run

```
(define (foo x y) (print x))
(foo "Hello, world!")
```

Results in

```
Hello, world!
```

Instead, it should give an function arity error.

**Comment by rain-1** [comment]
```
From e51024a3e04935ddceb90a1b0823def6c1e4fe55 Mon Sep 17 00:00:00 2001
From: rain1 <rain1@airmail.cc>
Date: Sun, 24 Mar 2019 20:14:41 +0000
Subject: [PATCH] Add a way to error when a lambda is called with the wrong
 number of arguments


diff --git a/src/system/compiler.js b/src/system/compiler.js
index 94e0916..a875133 100644
--- a/src/system/compiler.js
+++ b/src/system/compiler.js
@@ -487,6 +487,7 @@ BiwaScheme.Compiler = BiwaScheme.Class.create({
                     f.set_union(proper.to_set()),
                     ["return"]);
     var do_close = ["close", 
+                     vars.length(),
                      free.size(),
                      this.make_boxes(sets, proper, do_body),
                      next,
diff --git a/src/system/interpreter.js b/src/system/interpreter.js
index 2e1dc52..309c954 100644
--- a/src/system/interpreter.js
+++ b/src/system/interpreter.js
@@ -119,7 +119,7 @@ BiwaScheme.Interpreter = BiwaScheme.Class.create({
 
   // private
   //ret: [body, stack[s-1], stack[s-2], .., stack[s-n], dotpos]
-  closure: function(body, n, s, dotpos){
+  closure: function(body, n_args, n, s, dotpos){
     var v = []; //(make-vector n+1+1)
     v[0] = body;
     for(var i=0; i<n; i++)
@@ -127,6 +127,9 @@ BiwaScheme.Interpreter = BiwaScheme.Class.create({
     v[n+1] = dotpos;
 
     v.closure_p = true;
+    if(dotpos == -1) {
+      v.expected_args = n_args;
+    }
 
     return v;
   },
@@ -202,8 +205,8 @@ BiwaScheme.Interpreter = BiwaScheme.Class.create({
         break;
       case "close":
         var ox=x;
-        var n=ox[1], body=ox[2], x=ox[3], dotpos=ox[4];
-        a = this.closure(body, n, s, dotpos);
+        var v=ox[1], n=ox[2], body=ox[3], x=ox[4], dotpos=ox[5];
+        a = this.closure(body, v, n, s, dotpos);
         s -= n;
         break;
       case "box":
@@ -281,7 +284,7 @@ BiwaScheme.Interpreter = BiwaScheme.Class.create({
         // the number of arguments in the last call is
         // pushed to the stack.
         var n_args = this.index(s, -1);
-        if(func instanceof Array){ //closure
+        if(func instanceof Array && func.closure_p === true){ //closure
           a = func;
           x = func[0];
 
@@ -311,6 +314,15 @@ BiwaScheme.Interpreter = BiwaScheme.Class.create({
             }
             this.index_set(s, dotpos, ls);
           }
+          else {
+            // the dot is not found
+            // --------------------
+            // => Verify that number of arguments = expected number of arguments
+            // (if the closure knows how many it wants)
+            if(func.expected_args && n_args != func.expected_args) {
+              alert("Function call error: got " + n_args + " but wanted " + func.expected_args)
+            }
+          }
           f = s;
           c = a;
         }
-- 
2.21.0
```

**Comment by rain-1** [comment]
also `doc/internal/opecodes.md`, `["close", <n_args>, <n>, <body>, <opecodes_next>, <dotpos>]`

**Comment by lilsweetcaligula** [comment]
@rain-1 If you have prepared a fix, could you please submit a PR?

**Comment by rain-1** [comment]
It's not a full fix because I just use alert, I don't know the way to throw an exception and stop execution

**Comment by lilsweetcaligula** [comment]
@rain-1 In order to throw an exception and stop execution you may try the following:
```
var errMsg = 'errMsgPlaceholder';
throw new BiwaScheme.Error(errMsg);
```

**Comment by lilsweetcaligula** [comment]
@rain-1 Yeah, sure. You would probably want to write tests for this feature in `test/unit.js`.

In order to test how it works in REPL, you could use `demo/repl.html`.

**Comment by rain-1** [comment]
ah thank you, with `test/spec.html` I see my patch breaks 4 tests. I will investigate this.

**Comment by rain-1** [comment]
Fixed all the tests except this one:

```
define-macro
[BUG] you cannot use free variables in macro expander (or define-macro must be on toplevel) []

at undefined, line undefined
```

**Comment by rain-1** [comment]
Now we would like to add tests that verify that functions cannot be called with the wrong number of arguments. This patch has only done this in the case without a dot (without varargs) though.

**Comment by yhara** [comment]
> Fixed all the tests except this one:

Hmm, at first it looked easy to fix, but this patch causes another error 🤔 

```diff
--- a/src/library/extra_lib.js
+++ b/src/library/extra_lib.js
@@ -238,11 +238,11 @@ if( typeof(BiwaScheme)!='object' ) BiwaScheme={}; with(BiwaScheme) {
       expected_args = lambda.cdr.car;
     }

-    //[close, n_frees, do_body, next]
+    //[close, n_args, n_frees, do_body, next]
     var opc = Compiler.compile(lambda);
-    if(opc[1] != 0)
+    if(opc[2] != 0)
       throw new Bug("you cannot use free variables in macro expander (or define-macro must be on toplevel)")
-    var cls = [opc[2]];
+    var cls = [opc[3]];

     TopEnv[name.name] = new Syntax(name.name, function(sexp){
       var given_args = sexp.to_array();
```

**Comment by rain-1** [comment]
yeah this one is a difficult problem, i'm unable to use macros in my modified version.

**Comment by yhara** [comment]
@rain-1 All tests are passed by adding `cls.closure_p = true` and I've merged #132 (via #151). Thank you for your contribution!
